crates/ntx-builder/src/actor/mod.rs

```diff
@@ -0,0 +1,339 @@
+                    }
+                },
+                // Execute transactions.
+                permit = tx_permit_acquisition => {
+                    match permit {
+                        Ok(_permit) => {
+                            // Read the chain state.
+                            let chain_state = self.chain_state.read().await.clone();
+                            // Find a candidate transaction and execute it.
+                            if let Some(tx_candidate) = state.select_candidate(crate::MAX_NOTES_PER_TX, chain_state) {
+                                self.execute_transactions(&mut state, tx_candidate).await;
+                            } else {
+                                // No transactions to execute, wait for events.
+                                self.mode = ActorMode::NoViableNotes;
+                            }
+                        }
+                        Err(err) => {
+                            return ActorShutdownReason::SemaphoreFailed(err);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Execute a transaction candidate and mark notes as failed as required.
+    ///
+    /// Updates the state of the actor based on the execution result.
+    #[tracing::instrument(name = "ntx.actor.execute_transactions", skip(self, state, tx_candidate))]
+    async fn execute_transactions(
+        &mut self,
+        state: &mut NetworkAccountState,
+        tx_candidate: TransactionCandidate,
+    ) {
+        let block_num = tx_candidate.chain_tip_header.block_num();
+
+        // Execute the selected transaction.
+        let context = execute::NtxContext {
+            block_producer: self.block_producer.clone(),
+            prover: self.prover.clone(),
+        };
+
+        let execution_result = context.execute_transaction(tx_candidate).await;
+        match execution_result {
+            // Execution completed without failed notes.
+            Ok((tx_id, failed)) if failed.is_empty() => {
+                self.mode = ActorMode::TransactionInflight(tx_id);
+            },
+            // Execution completed with some failed notes.
+            Ok((tx_id, failed)) => {
+                let notes = failed.into_iter().map(|note| note.note).collect::<Vec<_>>();
+                state.notes_failed(notes.as_slice(), block_num);
+                self.mode = ActorMode::TransactionInflight(tx_id);
+            },
+            // Transaction execution failed.
+            Err(err) => {
+                tracing::error!(err = err.as_report(), "network transaction failed");
+                match err {
+                    NtxError::AllNotesFailed(failed) => {
+                        let notes = failed.into_iter().map(|note| note.note).collect::<Vec<_>>();
+                        state.notes_failed(notes.as_slice(), block_num);
+                        self.mode = ActorMode::NoViableNotes;
+                    },
```

Comment:
IIUC, this would set the mode of the actor to `NoViableNotes` if we failed to execute the selected transaction - right? If so, wouldn't this ignore the notes that were not selected?

Basically, let's say the actor has 51 pending notes, when we select a candidate, we pick 50 notes for a transaction (because `MAX_NOTES_PER_TX == 50`), and then if this transaction fails, we'll be in a situation that we still have 1 note to try but the actor is marked as `NoViableNotes`.

Also, `MAX_NOTES_PER_TX = 50` seems a bit high. I don't remember the numbers exactly, but I think we wanted to make sure we can select a candidate in under 50 ms. would running a set of 50 notes through note consumabilty checker take around 50 ms? (I thought it would be more).